Question 3

1.

In [5]:
#print("test")
import pandas as pd
import math
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
data=pd.read_csv(r"mammographic_masses.csv",delimiter=",",header=0)
#print(data)
d=data.values.tolist()
for each in range(len(d)):
    d[each][0]="BI-RADS:"+d[each][0]
    d[each][1] = "Age:" + d[each][1]
    d[each][2] = "Shape:" + d[each][2]
    d[each][3] = "Margin:" + d[each][3]
    d[each][4] = "Density:" + d[each][4]
    d[each][-1]="Severity:"+str(d[each][-1])
#print(d)
te = TransactionEncoder()
te_ary = te.fit(d).transform(d)
#print(te.columns_)
df = pd.DataFrame(te_ary,columns=te.columns_)
#print(df)

#computing frequent itemsets and association rules
frequent_itemsets = apriori(df, min_support=0.2, use_colnames=True)

a=association_rules(frequent_itemsets, metric="confidence", min_threshold=0.9)

#visualizing association rules results
print(a[["antecedents","consequents","support","confidence"]])

                          antecedents   consequents   support  confidence
0                           (Shape:4)   (Density:3)  0.374610    0.900000
1               (Density:3, Margin:1)   (BI-RADS:4)  0.263267    0.900356
2              (Severity:0, Margin:1)   (BI-RADS:4)  0.299688    0.911392
3               (BI-RADS:4, Margin:1)  (Severity:0)  0.299688    0.911392
4                (BI-RADS:5, Shape:4)   (Density:3)  0.245578    0.904215
5                (BI-RADS:5, Shape:4)  (Severity:1)  0.246618    0.908046
6               (Shape:4, Severity:1)   (Density:3)  0.295525    0.901587
7   (Severity:0, Density:3, Margin:1)   (BI-RADS:4)  0.238293    0.927126
8    (BI-RADS:4, Density:3, Margin:1)  (Severity:0)  0.238293    0.905138
9     (BI-RADS:5, Shape:4, Density:3)  (Severity:1)  0.224766    0.915254
10   (BI-RADS:5, Shape:4, Severity:1)   (Density:3)  0.224766    0.911392


As shown above, there is the result of the question 1, which shows the rules with support at least 0.2 and confidence at least 0.9.

2.

In [6]:
dnew=[]
for each in range(len(d)):
    dnew.append(d[each][1:])
#print(dnew)
te_new = TransactionEncoder()
te_ary_new = te_new.fit(dnew).transform(dnew)
#print(te.columns_)
df_new = pd.DataFrame(te_ary_new,columns=te_new.columns_)
#print(df_new)
frequent_itemsets_new = apriori(df_new, min_support=0.1, use_colnames=True)
#print(frequent_itemsets_new)
a=association_rules(frequent_itemsets_new, metric="confidence", min_threshold=0.9)
#print(a[["antecedents","consequents","support","confidence"]])
#"Severity:1"in a["antecedents"][i] or "Severity:0"in a["antecedents"][i]or

for i in range(len(a["antecedents"])):
    if ("Severity:1" in a["consequents"][i] or "Severity:0" in a["consequents"][i]):
        print(a["antecedents"][i],":",a["consequents"][i])

frozenset({'Shape:2', 'Margin:1'}) : frozenset({'Severity:0'})
frozenset({'Shape:1', 'Density:3', 'Margin:1'}) : frozenset({'Severity:0'})


As shown above, shape, density and margin of the lesion can help to determine whether a lesion is benign or malign. The people with Shape:'2', Margin:'1' always result in the benign, and  the people with Shape:'1',Density:'3', Margin:'1' are always benign too. 

3.


In [11]:
#Question 3
special_itemsets = apriori(df, min_support=0.002, use_colnames=True)
s=association_rules(special_itemsets, metric="confidence", min_threshold=0)
#print(a)
for i in range(len(s)):
    if ("BI-RADS:6" in s["antecedents"][i] and "Severity:0" in s["consequents"][i]):
        print("Antecedents: ",s["antecedents"][i],",Consequents: ",s["consequents"][i])
for i in range(len(s)):
    if ("BI-RADS:0" in s["antecedents"][i] and "Severity:1" in s["consequents"][i]):
        print("Antecedents: ",s["antecedents"][i],",Consequents: ",s["consequents"][i])

Antecedents:  frozenset({'BI-RADS:6'}) ,Consequents:  frozenset({'Severity:0'})
Antecedents:  frozenset({'BI-RADS:6', 'Density:3'}) ,Consequents:  frozenset({'Severity:0'})
Antecedents:  frozenset({'BI-RADS:6'}) ,Consequents:  frozenset({'Severity:0', 'Density:3'})
Antecedents:  frozenset({'BI-RADS:6', 'Margin:1'}) ,Consequents:  frozenset({'Severity:0'})
Antecedents:  frozenset({'BI-RADS:6'}) ,Consequents:  frozenset({'Severity:0', 'Margin:1'})
Antecedents:  frozenset({'BI-RADS:6', 'Density:3', 'Margin:1'}) ,Consequents:  frozenset({'Severity:0'})
Antecedents:  frozenset({'BI-RADS:6', 'Density:3'}) ,Consequents:  frozenset({'Severity:0', 'Margin:1'})
Antecedents:  frozenset({'BI-RADS:6', 'Margin:1'}) ,Consequents:  frozenset({'Severity:0', 'Density:3'})
Antecedents:  frozenset({'BI-RADS:6'}) ,Consequents:  frozenset({'Severity:0', 'Density:3', 'Margin:1'})
Antecedents:  frozenset({'BI-RADS:0'}) ,Consequents:  frozenset({'Severity:1'})
Antecedents:  frozenset({'BI-RADS:0', 'Density:3'}

As shown above, there are several rules show that BI-RADS assessment is not accurate. When the BI-RADS is 6(the lesion is highly suggetive of malignancy), the severity is acturally 0, which means that the lesion is benign. And when Antecedents is BI-RADS:0(definitely benign), but the Consequents is Severity:1(malignant). These rules above prove that BI-RADS assessment sometimes can be wrong, and it is not always accurate.

To find the case which shows the BI-RADS assessment is not that accurate, I set the min_support and min_confidence as a value close to 0. Thus, I can find the counter example, which shows the BI-RADS assessment has the wrong result.

In the code above, I just detect the rules with Antecedents is BI-RADS:6 - Consequents is Severity:0 and Antecedents is BI-RADS:0 - Consequents is Severity:1. There are more rules with BI-RADS showing the lesion is very likely to be benign, but the Severity is 1(malignant). 


4.


In [13]:
#Question 4
frequent_itemsets = apriori(df, min_support=0.001, use_colnames=True)
a=association_rules(frequent_itemsets, metric="confidence", min_threshold=0)
for i in range(len(a)):
    if (a["antecedents"][i]==frozenset({'Age:35'}) and a["consequents"][i])==frozenset({'Severity:0'}):
        print(a["antecedents"][i],":",a["consequents"][i],",Support:",a["support"][i],",Confidence:",a["confidence"][i])


frozenset({'Age:35'}) : frozenset({'Severity:0'}) ,Support: 0.012486992715920915 ,Confidence: 0.923076923076923


For the rule: Age=35 ⇒ Severity=0, the confidence is 0.923076923076923, and the support is 0.012486992715920915. 

With the confidence is 0.923076923076923, which is higher than 0.9, I think this rule tell us something valuable. It show that people with age:35, their lesion is more likely to be benign.

5.

In [22]:
#Question 5
data=pd.read_csv(r"mammographic_masses.csv",delimiter=",",header=0)
#print(data)
#index_special=data[data['Age'].isin(["?"])].index.tolist()
#print(index_special)
#data_new=data.drop(index_special)#drop the rows with Age is ?
#give up the method above
d_num=data.values.tolist()
sum_age=0
for each in range(len(d_num)):
    d_num[each][0]="BI-RADS:"+d_num[each][0]
    if d_num[each][1]!="?":
        d_num[each][1] = int(d_num[each][1])
        sum_age+=d_num[each][1]
    d_num[each][2] = "Shape:" + d_num[each][2]
    d_num[each][3] = "Margin:" + d_num[each][3]
    d_num[each][4] = "Density:" + d_num[each][4]
    d_num[each][-1]="Severity:"+str(d_num[each][-1])
mean=sum_age/len(d_num)
mean=int(mean)
#print(mean)
for each in range(len(d_num)):
    if d_num[each][1]=="?":
        d_num[each][1]="Age<="+str(mean)
    elif d_num[each][1]>mean:
        d_num[each][1] = "Age>"+str(mean)
        #print(d_num[each][1])
    else:
        d_num[each][1] = "Age<=" + str(mean)

te = TransactionEncoder()
te_ary = te.fit(d_num).transform(d_num)
#print(te.columns_)
df = pd.DataFrame(te_ary,columns=te.columns_)
frequent_itemsets = apriori(df, min_support=0.1, use_colnames=True)

a=association_rules(frequent_itemsets, metric="confidence", min_threshold=0.9)
#visualizing association rules results
print(a[["antecedents","consequents","support","confidence"]])

                                    antecedents   consequents   support  \
0                                    (Margin:5)   (Density:3)  0.131113   
1                                     (Shape:4)   (Density:3)  0.374610   
2                           (Age<=55, Margin:1)  (Severity:0)  0.229969   
3                            (Shape:1, Age<=55)  (Severity:0)  0.140479   
4                            (Shape:2, Age<=55)  (Severity:0)  0.113424   
..                                          ...           ...       ...   
63      (BI-RADS:5, Shape:4, Density:3, Age>55)  (Severity:1)  0.163371   
64     (BI-RADS:5, Severity:1, Shape:4, Age>55)   (Density:3)  0.163371   
65  (Severity:0, BI-RADS:4, Density:3, Shape:1)    (Margin:1)  0.133195   
66   (Severity:0, Margin:1, Density:3, Shape:1)   (BI-RADS:4)  0.133195   
67    (Shape:1, BI-RADS:4, Density:3, Margin:1)  (Severity:0)  0.133195   

    confidence  
0     0.926471  
1     0.900000  
2     0.917012  
3     0.924658  
4     0.900826

antecedents: (Age<=55, Margin:1) 

consequents: (Severity:0)

support: 0.229969 

confidence: 0.917012

(Age<=55, Margin:1)→(Severity:0)

The answer is shown above.